src/encrypt.rs

```diff
@@ -0,0 +1,464 @@
+
+use crate::error::COSEError;
+use crate::header_map::{map_to_empty_or_serialized, HeaderMap};
+
+#[allow(dead_code)]
+enum COSEAlgorithm {
+    /// AES-GCM mode w/ 128-bit key, 128-bit tag
+    A128GCM,
+    /// AES-GCM mode w/ 192-bit key, 128-bit tag
+    A192GCM,
+    /// AES-GCM mode w/ 256-bit key, 128-bit tag
+    A256GCM,
+    /// AES-CCM mode 128-bit key, 64-bit tag, 13-byte nonce
+    AesCcm16_64_128,
+    /// AES-CCM mode 256-bit key, 64-bit tag, 13-byte nonce
+    AesCcm16_64_256,
+    /// AES-CCM mode 128-bit key, 64-bit tag, 7-byte nonce
+    AesCcm64_64_128,
+    /// AES-CCM mode 256-bit key, 64-bit tag, 13-byte nonce
+    AesCcm64_64_256,
+    /// AES-CCM mode 128-bit key, 128-bit tag, 13-byte nonce
+    AesCcm16_128_128,
+    /// AES-CCM mode 256-bit key, 128-bit tag, 13-byte nonce
+    AesCcm16_128_256,
+    /// AES-CCM mode 128-bit key, 128-bit tag, 7-byte nonce
+    AesCcm64_128_128,
+    /// AES-CCM mode 128-bit key, 128-bit tag, 7-byte nonce
+    AesCcm64_128_256,
+}
+
+impl COSEAlgorithm {
+    fn value(&self) -> usize {
+        match self {
+            COSEAlgorithm::A128GCM => 1,
+            COSEAlgorithm::A192GCM => 2,
+            COSEAlgorithm::A256GCM => 3,
+            COSEAlgorithm::AesCcm16_64_128 => 10,
+            COSEAlgorithm::AesCcm16_64_256 => 11,
+            COSEAlgorithm::AesCcm64_64_128 => 12,
+            COSEAlgorithm::AesCcm64_64_256 => 13,
+            COSEAlgorithm::AesCcm16_128_128 => 30,
+            COSEAlgorithm::AesCcm16_128_256 => 31,
+            COSEAlgorithm::AesCcm64_128_128 => 32,
+            COSEAlgorithm::AesCcm64_128_256 => 33,
+        }
+    }
+
+    fn tag_size(&self) -> usize {
+        match self {
+            COSEAlgorithm::A128GCM => 16,
+            COSEAlgorithm::A192GCM => 16,
+            COSEAlgorithm::A256GCM => 16,
+            COSEAlgorithm::AesCcm16_64_128 => 8,
+            COSEAlgorithm::AesCcm16_64_256 => 8,
+            COSEAlgorithm::AesCcm64_64_128 => 8,
+            COSEAlgorithm::AesCcm64_64_256 => 8,
+            COSEAlgorithm::AesCcm16_128_128 => 16,
+            COSEAlgorithm::AesCcm16_128_256 => 16,
+            COSEAlgorithm::AesCcm64_128_128 => 16,
+            COSEAlgorithm::AesCcm64_128_256 => 16,
+        }
+    }
+
+    fn openssl_cipher(value: i8) -> Option<Cipher> {
+        Some(match value {
+            1 => Cipher::aes_128_gcm(),
+            2 => Cipher::aes_192_gcm(),
+            3 => Cipher::aes_256_gcm(),
+            30 => Cipher::aes_128_ccm(),
+            31 => Cipher::aes_256_ccm(),
+            _ => return None,
+        })
+    }
+
+    fn from_nid(nid: openssl::nid::Nid) -> Option<COSEAlgorithm> {
+        Some(match nid {
+            openssl::nid::Nid::AES_128_GCM => COSEAlgorithm::A128GCM,
+            openssl::nid::Nid::AES_192_GCM => COSEAlgorithm::A192GCM,
+            openssl::nid::Nid::AES_256_GCM => COSEAlgorithm::A256GCM,
+            openssl::nid::Nid::AES_128_CCM => COSEAlgorithm::AesCcm16_128_128,
```

Comment:
Wouldn't this Nid  fit both  `AesCcm16_128_128` and `AesCcm64_128_128`? Same for the others.

Reply:
uhm, not sure, @puiterwijk ?

Reply:
It would be ideal to not surface OpenSSL-specific structures if possible, since I was considering adding another crypto backend and it would reduce the number of API changes if this parameter was contained to this library.

Reply:
dropped CCM and implemented a `CipherConfiguration` enum to take care of not exposing openssl directly